### ISSUE TITLE
performance impact of auto `automount: true`

### ISSUE THREAD

**Issue by phillipskevin** [canjs/can-component#194]
TLDR: `autoMount`ing by default would add about 1ms of overhead for every 500 elements added to the page.

I ran some tests to estimate the performance impact of making `autoMount: true` happen automatically. This would work by creating a MutationObserver that would listen for added nodes and run the automount code: 

https://github.com/canjs/can-component/blob/ecdadc08e7c6c1722fc4a3b9500e66a59e8ac7ca/can-component.js#L227-L239

Here is the code I used to test this: https://gist.github.com/phillipskevin/5e1cd05b92960c0aa09e485d55ca8f1c

And here are some screenshots showing the time spent in the `MUTATION_HANDLER` function. I ran the tests for 500, 5000, and 50000 elements and also tests with these same number of elements, but split across 5 and 10 different calls to `appendChild` (for example the final test was inserting 5000 nodes 10 times, for a total of 50000 nodes).

![image](https://user-images.githubusercontent.com/5851984/34694482-e2341df2-f48c-11e7-99a2-3c44094abfb0.png)

![image](https://user-images.githubusercontent.com/5851984/34694489-e4fe87f2-f48c-11e7-98aa-a509d7581532.png)

![image](https://user-images.githubusercontent.com/5851984/34694491-e7c08620-f48c-11e7-89c4-1caf2ce95269.png)

![image](https://user-images.githubusercontent.com/5851984/34694493-eac3fee2-f48c-11e7-819a-9d553f8e9a02.png)

![image](https://user-images.githubusercontent.com/5851984/34694497-ed907e3e-f48c-11e7-9843-82b717f9a0f6.png)

![image](https://user-images.githubusercontent.com/5851984/34694501-f01c0998-f48c-11e7-99c0-5284070316b1.png)

![image](https://user-images.githubusercontent.com/5851984/34694505-f2dfb7a6-f48c-11e7-9ccc-55d4d57781a2.png)

![image](https://user-images.githubusercontent.com/5851984/34694509-f58fdbc0-f48c-11e7-841d-8ddbb011b607.png)

![image](https://user-images.githubusercontent.com/5851984/34694511-f7c46a32-f48c-11e7-85c4-c03d24a5f3f7.png)

**Comment by phillipskevin** [comment]
Summary of Results in Chrome:

Elements | Total Time in MUTATION_HANDLER | Time / 500 Elements
--------- | ------------------------------ | ---------------
500 | 0.9ms | 0.9ms |
100 * 5 | 0.8ms | 0.8ms |
50 * 10 | 1.0ms | 1.0ms |
5000 | 10.1ms | 1.01ms |
1000 * 5 | 13.6ms | 1.36ms |
500 * 10 |10.1ms | 1.01ms |
50000 | 120.2ms | 1.20ms |
10000 * 5 | 106.5ms |1.06ms  |
5000 * 10 | 96.0ms | 0.96ms |

**Comment by phillipskevin** [comment]
Summary of Results in Firefox:

(I used the same approach as Chrome - ran the test and looked the time in the Call Tree in the Performance Profiler in Firefox devtools)

Elements | Total Time in MUTATION_HANDLER | Time / 500 Elements
--------- | ------------------------------ | ---------------
500 | - | - |
100 * 5 | - | - |
50 * 10 | -  | - |
5000 | 1.04ms | 0.10ms |
1000 * 5 | 1.06ms  | 0.10ms |
500 * 10 | 1.07ms | 0.10ms |
50000 | 15.03ms  | 0.15ms |
10000 * 5 | 17.06ms  | 0.17ms  |
5000 * 10 | 15.23ms | 0.15ms |

**Comment by phillipskevin** [comment]
Summary of Results in Safari:

(I used the same approach as Chrome - ran the test and looked the time in the JavaScript & Events section in theTimelines tab in Safari devtools)

Elements | Total Time in MUTATION_HANDLER | Time / 500 Elements
--------- | ------------------------------ | ---------------
500 | 0.662ms | 0.662ms |
100 * 5 | 0.694ms | 0.694ms |
50 * 10 | - | - |
5000 | 1.909ms | 0.19ms |
1000 * 5 |  2.147ms | 0.21ms |
500 * 10 | 2.396ms | 0.24ms |
50000 | - | - |
10000 * 5 | - | - |
5000 * 10 | - | - |
  
Results showing `-` did not register the MUTATION_HANDLER at all.

**Comment by phillipskevin** [comment]
In Edge, I could not get any set of tests to register the MUTATION_HANDLER in call stack of the performance profiler. I'm taking this to mean it is a very small amount of the overall time. I tried with a SauceLabs VM and an actual Device (MS Surface).

**Comment by justinbmeyer** [comment]
@phillipskevin I'm not sure this is the best way of profiling.  Could you measure total time with and without and subtract?

**Comment by phillipskevin** [comment]
I could do that, but running the same test repeatedly can have pretty large differences in total time:

![image](https://user-images.githubusercontent.com/5851984/34739763-f26d83b4-f542-11e7-8d79-cf818de506b4.png)

If the test with the MutationObserver is 25ms longer than the test without, how do I know that it is the MutationObserver that caused it and not just normal variance? I would have to run the test many times to get significant data.

That's why I thought using the profiler to see exactly how much time was spent in the handler was the way to go.

**Comment by phillipskevin** [comment]
Based on this, we're going to go ahead with these changes. I'm going to submit a new issue with the details.
